server/historian/packages/historian-base/src/routes/utils.ts

```diff
@@ -47,9 +62,147 @@ export class createGitServiceArgs {
 }
 
+function getEphemeralContainerCacheKey(documentId: string): string {
+	return `isEphemeralContainer:${documentId}`;
+}
+
+async function updateIsEphemeralCache(
+	documentId: string,
+	tenantId: string,
+	isEphemeral: boolean,
+	cache: ICache,
+): Promise<void> {
+	const isEphemeralKey: string = getEphemeralContainerCacheKey(documentId);
+	Lumberjack.info(
+		`Setting cache for ${isEphemeralKey} to ${isEphemeral}.`,
+		getLumberBaseProperties(documentId, tenantId),
+	);
+	await runWithRetry(
+		async () => cache?.set(isEphemeralKey, isEphemeral) /* api */,
+		"utils.createGitService.set" /* callName */,
+		3 /* maxRetries */,
+		1000 /* retryAfterMs */,
+		getLumberBaseProperties(documentId, tenantId) /* telemetryProperties */,
+	).catch((error) => {
+		Lumberjack.error(
+			`Error setting ${isEphemeralKey} in redis`,
+			getLumberBaseProperties(documentId, tenantId),
+			error,
+		);
+	});
+}
+
+/**
+ * Checks the cache for the isEphemeral flag for the given document. If the flag is not in the cache, it fetches the
+ * static data from the document and caches the flag.
+ *
+ * Returns false if the flag is not found in the static data will cause Ephemeral Container access
+ * to fail, which is better than returning true and causing durable container access to fail.
+ * Ephemeral containers do not have as strong of an SLA as durable containers, so this is the safer choice.
+ *
+ * @returns - Returns a boolean indicating whether the container is ephemeral or not.
+ * @throws - Throws an error if falling back to static data indicates an ephemeral document is older than the provided
+ * max ephemeral document TTL.
+ */
+async function checkAndCacheIsEphemeral({
+	documentId,
+	tenantId,
+	documentManager,
+	ephemeralDocumentTTLSec,
+	isEphemeralContainerOverride,
+	cache,
+}: {
+	documentId: string;
+	tenantId: string;
+	documentManager: IDocumentManager;
+	ephemeralDocumentTTLSec: number;
+	isEphemeralContainerOverride?: boolean;
+	cache?: ICache;
+}): Promise<boolean> {
+	const isEphemeralKey = getEphemeralContainerCacheKey(documentId);
+	if (typeof isEphemeralContainerOverride === "boolean") {
+		// If an isEphemeralContainerOverride flag was passed in, cache it and return it.
+		// This typically happens on the first request to the document.
+		await updateIsEphemeralCache(documentId, tenantId, isEphemeralContainerOverride, cache);
+		return isEphemeralContainerOverride;
+	}
+	// When no override is provided, check the cache first.
+	const cachedIsEphemeral: boolean | undefined = await runWithRetry<boolean | undefined>(
+		async () => cache?.get(isEphemeralKey) /* api */,
+		"utils.createGitService.get" /* callName */,
+		3 /* maxRetries */,
+		1000 /* retryAfterMs */,
+		getLumberBaseProperties(documentId, tenantId) /* telemetryProperties */,
+	).catch((error) => {
+		Lumberjack.error(
+			`Error getting ${isEphemeralKey} from redis`,
+			getLumberBaseProperties(documentId, tenantId),
+			error,
+		);
+		return undefined;
```

Comment:
Don't we want to attempt to fetch the flag from db when we get an exception while checking for the flag in the cache?

Reply:
Yes, this is in the `.catch` for the cache flag. If the flag returned is `undefined` (e.g. on error), then we will fallback to DB.

Reply:
I'll switch some of these to try..catch statements to make it more clear.